Add Edools::LessonProgress

herminiotorres opened the pull request:
- Add an error logger.
- Add endpoint to LessonProgress with pagination if necessary.
- Add example in README.
- Version bump.

herminiotorres committed:
Add Edools::LessonProgress

- Add an error logger.
- Add endpoint to LessonProgress with pagination if necessary.
- Add example in README.
- Version bump.

regismesquita commented on README.md:
````diff
@@ -74,6 +74,42 @@ Edools::Enrollment.update(id: 1, school_product_id: 75506, status: "active")
 Edools::Enrollment.all
 ```
 
+### LessonProgress example
````
Please add the expected return format at the readme as the openstruct can be a bit surprising.

regismesquita commented on lib/edools/lesson_progress.rb:
```diff
@@ -0,0 +1,148 @@
+
+      response = Edools::ApiRequest.request(:get, url, data)
+
+      OpenStruct.new(
```
Maybe we could return something like a "collectionProxy" here , allowing the user a more straightforward approach like `Edools::LessonProgresses.find_paginated_from_enrollments.first` ? I think that we can ship it with your current structure as it is easy to keep retro-compatibility and add an issue for providing CollectionProxy-like capabilities in the future.

herminiotorres replied:
the collection looks like this:
`Edools::LessonProgresses.find_paginated_from_enrollments.lessons_progresses.first`

regismesquita replied:
Exactly , that behavior is not what one would normally expect hence the request to make clear the return definition inside the README, it's not common the need to refer to what object you want right after doing a find on the class that define that object.

A CollectionProxy-like would allow us to retain the expected behaviour (i.e. `Edools::LessonProgresses.find_paginated_from_enrollments.first` ) and at the same time to refer to pagination related information.

But as stated before we can ship it and change that later.

regismesquita commented on lib/edools/lesson_progress.rb:
```diff
@@ -0,0 +1,148 @@
+    private
+
+    def self.objectify_paginate(paginate)
+      return paginate if paginate.nil?
```
When paginate is going to be `nil`? If that happens when there's only a single page I imagine users may rely on this pagination information in order to avoid multiples calls, so maybe we can generate a structure like the one below just to keep the standard?

something like: 
```
{
  current_page: 1,
  per_page: lessons_progresses.size,
  total_pages: 1,
  total_count: lessons_progresses.size
}
```

Does it make sense? I fell like this is a much more expected behaviour although being able to `unless foobar.paginate` seems cool too.

herminiotorres replied:
I prevent any possible problem that the pagination will have, so I do this nil validation.

I'm more comfortable passing the values coming from the API

regismesquita commented on lib/edools/lesson_progress.rb:
```diff
@@ -0,0 +1,148 @@
+    end
+
+    def self.objectify_lesson_progress(lesson_progress)
+      return lesson_progress if lesson_progress.nil?
```
If we can have `nil` inside the array maybe we should compact the array created inside `self.objectify_lessons_progresses`?

beneti commented on README.md:
````diff
@@ -74,6 +74,42 @@ Edools::Enrollment.update(id: 1, school_product_id: 75506, status: "active")
+
+Using `enrollment_id` as a parameter to search for LessonProgress:
+```ruby
+Edools::LessonProgress.find_paginated_from_enrollments(enrollment_id: 2075731)
````
I would use

`find_paginated_by_enrollment`, in singular, you can't pass an array for example, and `by`, to be rails like.

regismesquita replied:
Agree.

beneti commented on README.md:
````diff
@@ -74,6 +74,42 @@ Edools::Enrollment.update(id: 1, school_product_id: 75506, status: "active")
+
+Using `student_id` as a parameter to search for LessonProgress:
+```ruby
+Edools::LessonProgress.find_paginated_from_students(student_id: 6558938)
````
I would use

`find_paginated_by_student`, in singular, you can't pass an array for example, and `by`, to be rails like.

regismesquita replied:
Agree.

beneti commented on README.md:
````diff
@@ -74,6 +74,42 @@ Edools::Enrollment.update(id: 1, school_product_id: 75506, status: "active")
+
+Using `school_product_id` as a parameter to search for LessonProgress:
+```ruby
+Edools::LessonProgress.find_paginated_from_school_products(school_product_id: 73636)
````
I would use

`find_paginated_by_school_product`, in singular, you can't pass an array for example, and `by`, to be rails like.

regismesquita replied:
Agree.

beneti commented on README.md:
````diff
@@ -74,6 +74,42 @@ Edools::Enrollment.update(id: 1, school_product_id: 75506, status: "active")
+
+Other parameters that can be passed to the endpoint:
+```ruby
+result = Edools::LessonProgress.find_paginated_from_school_products(school_product_id: 73636, page: 2, per_page: 30)
````
in final, I think I would use

`find_paginated_by`, and handle if it is school_prduct, student or enrollment in this method

regismesquita replied:
Agree.

beneti commented on lib/edools/lesson_progress.rb:
```diff
@@ -0,0 +1,148 @@
+      @errors = data.dig(:errors)
+    end
+
+    def self.find_paginated_from_enrollments(enrollment_id: , page: 1, per_page: 50)
```
personally (I don't know if I can use this here lol), I don't like magic numbers.

I would use `DEFAULT_PAGE` and `DEFAULT_PER_PAGE`

regismesquita replied:
Makes sense and won't hurt, also next to the constant definition we can add comments explaining why such numbers were used.

herminiotorres committed:
Improvements in Edools::LessonProgress

- Add the OpenStruct return format in the README.
- Compact lesson progress if nil.
- Rename method from find_paginated_from to find_paginated_by.
- Assign the default page and per page values to static class variables, avoid magic numbers.

herminiotorres committed:
Add a PaginationProxy
